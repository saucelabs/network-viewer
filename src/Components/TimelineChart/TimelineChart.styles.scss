@import "./../../styles/variables";

.chart-container {
  margin: 0 $size-xs * 2;
}

.scatter-chart {
  tspan {
    font-size: 10px;
    fill: $white-50;
  }
}
