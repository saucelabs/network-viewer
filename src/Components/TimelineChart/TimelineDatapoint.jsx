import React, { useMemo } from 'react';
import PropTypes from 'prop-types';

import { calcChartAttributes } from './../../utils';
import { TIMELINE_DATA_POINT_HEIGHT } from '../../constants';

const TimelineDatapoint = ({ payload, maxTime, cx, cy, index }) => {
  const { timings } = payload;
  const chartAttributes = useMemo(
    () => calcChartAttributes(timings, maxTime, cx, index, cy), [timings, maxTime],
  );
  if (!payload) {
    return null;
  }

  return (
    <g>
      {chartAttributes.map((chartProps) => (
        <rect
          key={chartProps.key}
          {...chartProps}
          height={TIMELINE_DATA_POINT_HEIGHT}
        />
      ))}
    </g>
  );
};

TimelineDatapoint.propTypes = {
  cx: PropTypes.number,
  cy: PropTypes.number,
  index: PropTypes.number,
  maxTime: PropTypes.number.isRequired,
  payload: PropTypes.object,
};

TimelineDatapoint.defaultProps = {
  cx: 0,
  cy: 0,
  index: 0,
  payload: null,
};

export default TimelineDatapoint;
