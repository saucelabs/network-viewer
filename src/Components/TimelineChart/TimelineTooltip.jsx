import React from 'react';
import PropTypes from 'prop-types';

import Styles from './TimelineTooltip.styles.scss';
import { formatTime } from '../../utils';

const TimelineTooltip = ({ payload }) => {
  if (!payload || !payload.length) {
    return null;
  }

  return (
    <div className={Styles.tooltip}>
      <div className={Styles.content}>
        {payload[0].payload.filename}
      </div>
      <div>
        {`Started at: ${formatTime(payload[0].payload.timings.startTime)}`}
      </div>
    </div>
  );
};

TimelineTooltip.propTypes = {
  payload: PropTypes.array,
};

TimelineTooltip.defaultProps = {
  payload: null,
};

export default TimelineTooltip;
