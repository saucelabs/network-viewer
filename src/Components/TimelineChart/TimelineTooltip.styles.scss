@import "./../../styles/variables.scss";

.tooltip {
  display: block;
  width: 300px;
  height: auto;
  background: $white-100;
  border: 1px solid $white-90;
  border-radius: 3px;
  color: $white-33;
  font-size: $font-size-small;
  padding: $size-m / 2;
  opacity: 1;
}

.content {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}
