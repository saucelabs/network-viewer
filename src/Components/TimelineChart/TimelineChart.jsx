import React from 'react';
import PropTypes from 'prop-types';
import { ScatterChart, Scatter, XAxis, YAxis, ResponsiveContainer, Tooltip } from 'recharts';

import Styles from './TimelineChart.styles.scss';
import TimelineDatapoint from './TimelineDatapoint';
import { formatTime } from '../../utils';
import TimelineTooltip from './TimelineTooltip';

const TimelineChart = ({ chartData, totalNetworkTime }) => (
  <div className={Styles['chart-container']}>
    <ResponsiveContainer
      height={100}
      width="100%"
    >
      <ScatterChart
        className={Styles['scatter-chart']}
      >
        <XAxis
          axisLine={false}
          dataKey="timings.startTime"
          domain={[0, totalNetworkTime]}
          interval="preserveStartEnd"
          orientation="top"
          tickCount={10}
          tickFormatter={formatTime}
          tickLine={false}
          type="number"
        />
        <YAxis
          dataKey="index"
          domain={['min', 'max']}
          hide
          reversed
        />
        <Tooltip content={<TimelineTooltip />} />
        <Scatter
          data={chartData}
          shape={(
            <TimelineDatapoint
              maxTime={totalNetworkTime}
            />
          )}
        />
      </ScatterChart>
    </ResponsiveContainer>
  </div>
);

TimelineChart.propTypes = {
  chartData: PropTypes.array.isRequired,
  totalNetworkTime: PropTypes.number.isRequired,
};

export default TimelineChart;
