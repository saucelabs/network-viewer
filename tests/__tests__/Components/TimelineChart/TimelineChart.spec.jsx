import React from 'react';
import { mount } from 'enzyme';

import TimelineChart from './../../../../src/Components/TimelineChart/TimelineChart';

describe('TimelineChart', () => {
  it('renders without crashing', () => {
    const element = mount(<TimelineChart
      chartData={[]}
      totalNetworkTime={17000}
    />);
    expect(element).toMatchSnapshot();
  });
});
