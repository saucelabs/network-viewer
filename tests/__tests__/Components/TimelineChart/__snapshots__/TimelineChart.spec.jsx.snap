// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimelineChart renders without crashing 1`] = `
<TimelineChart
  chartData={Array []}
  totalNetworkTime={17000}
>
  <div>
    <ResponsiveContainer
      debounce={0}
      height={100}
      width="100%"
    >
      <div
        className="recharts-responsive-container"
        style={
          Object {
            "height": 100,
            "maxHeight": undefined,
            "minHeight": undefined,
            "minWidth": undefined,
            "width": "100%",
          }
        }
      >
        <ScatterChart
          barCategoryGap="10%"
          barGap={4}
          height={100}
          layout="horizontal"
          margin={
            Object {
              "bottom": 5,
              "left": 5,
              "right": 5,
              "top": 5,
            }
          }
          reverseStackOrder={false}
          stackOffset="none"
          width={0}
        />
        <ResizeDetector
          handleHeight={true}
          handleWidth={true}
          onResize={[Function]}
          refreshRate={1000}
          resizableElementId=""
          skipOnMount={false}
        >
          <div
            style={
              Object {
                "display": "none",
                "height": 0,
                "position": "absolute",
                "visibility": "hidden",
                "width": 0,
              }
            }
          />
        </ResizeDetector>
      </div>
    </ResponsiveContainer>
  </div>
</TimelineChart>
`;
